### ISSUE TITLE
Resource loaded without being associated with script class name

### ISSUE THREAD

**Issue by blackears** [godotengine/godot#87397]
### Tested versions

v4.3.dev2.official [352434668]

### System information

Windows 10 Mobile

### Issue description

I'm running into a strange error where attempting to load a custom resource with a `class_name` results in an object with none of the resources properties and which cannot be passed to any method which requires the `class_name`.  

It seems to be related to having a resource that contains an array that preloads other resources.  In my original project, I was using custom resources as UIDs to reference game levels and items.  I wanted to create a class that I could use as an index of all rooms and items in my game which my game's save file loader could reference to turn string ids back into these resource UIDs.  However, this seems to be breaking things.

### Steps to reproduce

Run the attached project.  Press the Test button.

### Minimal reproduction project (MRP)

[resource_load_test.zip](https://github.com/godotengine/godot/files/13997200/resource_load_test.zip)

**Comment by bs-mwoerner** [comment]
Possibly related:
- #87019

**Comment by blackears** [comment]
I've reworked the code so that the lists just contain String names and I `load()` them later.  This works around the issue.  And I didn't realize this `preload` this was likely what was causing the issue until I was finishing up this bug report.  But Godot's inability to handle cyclic references makes any use of `preload` or `PackedScene` an obscure bug just waiting to happen.  It can be really hard to make sure you're not accidentally creating a cycle when you're using custom resources and your code starts to get large.

**Comment by blackears** [comment]
Yes, this looks pretty similar to that other error you referenced.

**Comment by Maran23** [comment]
I got a similar error but even without preloading.
But otherwise the same situation as you have.
I have a Dictionary with object in it. And for some he does not detect what type of Resource it is, therefore I always get an error.
`Invalid call. Nonexistent function 'xxx' in base 'Resource'`.

EDIT: Culprit seems to be that I reference the `Resource` somewhere, which might be autoloaded.
In my case, it is more complex so your example is much better.

### Still for debugging, my scenario:
`Player` is autoloaded -> It references a Node called `Formation` which references a Resource called `Item` that references a Resource `ItemInfo`. `ItemInfo` can create an `Item` (so references `Item` indirectly).
Because of this, later when I do an item lookup, it will fail with the error above and it looks like Godot is not detecting that the `Resource` is of type `ItemInfo`.
If I removed the typing in `Item`, e.g. from `get_info() -> ItemInfo` to `get_info() -> Variant`, it works.

It seems to be a cyclic issue once again, as:
Before:
`Player` -> `Formation` -> `Item` -> `ItemInfo` -> `Item`

Now:
`Player` -> `Formation` -> `Item` (-> `ItemInfo` is a `Variant`)

**Comment by Viele** [comment]
I just had this issue with an Autoload script.
After writing it the game would randomly fail loading a specific resource.
The fix was to remove all type info from that script i.e. I had `Array[Item]` but i had to remove the `Item`

Disabling the script from autoload did not help.
